aptos-move/framework/aptos-framework/sources/HashMap.move

```diff
@@ -0,0 +1,291 @@
+    /// Destroy non-empty hashmap.
+    const ENOT_EMPTY: u64 = 2;
+
+    /// HashMap entry contains both the key and value.
+    struct Entry<K, V> has store {
+        hash: u64,
+        key: K,
+        value: V,
+    }
+
+    /// A hashmap implementation based on linear hashing. (https://www.cs.cmu.edu/afs/cs.cmu.edu/user/christos/www/courses/826-resources/PAPERS+BOOK/linear-hashing.PDF)
+    /// Compare to Table, it uses less storage slots but has higher chance of collision, it's a trade-off between space and time.
+    /// Compare to other hash map implementation, linear hashing splits one bucket a time instead of doubling buckets when expanding to avoid unexpected gas cost.
+    struct HashMap<K, V> has store {
+        buckets: Table<u64, vector<Entry<K, V>>>,
+        num_buckets: u64,
+        // number of bits to represent num_buckets
+        level: u8,
+        // total number of items
+        len: u64,
+    }
+
+    /// Create an empty HashMap with `initial_buckets` buckets.
+    public fun new<K: drop + store, V: store>(initial_buckets: u64): HashMap<K, V> {
+        assert!(initial_buckets > 0 && (initial_buckets & (initial_buckets - 1) == 0), Errors::invalid_argument(ENOT_POWER_OF_TWO));
+        let buckets = Table::new();
+        let i = 0;
+        let level = 0;
+        while (i < initial_buckets) {
+            Table::add(&mut buckets, &i, Vector::empty());
+            i = i + 1;
+            if (i == 1 << (level + 1)) {
+                level = level + 1;
+            }
+        };
+        HashMap {
+            buckets,
+            num_buckets: initial_buckets,
+            level,
+            len: 0,
+        }
+    }
+
+    /// Destroy empty map.
+    /// Aborts if it's not empty.
+    public fun destroy_empty<K, V>(map: HashMap<K, V>) {
+        assert!(map.len == 0, Errors::invalid_argument(ENOT_EMPTY));
+        let i = 0;
+        while (i < map.num_buckets) {
+            Vector::destroy_empty(Table::remove(&mut map.buckets, &i));
+            i = i + 1;
+        };
+        let HashMap {buckets, num_buckets: _, level: _, len: _} = map;
+        Table::destroy_empty(buckets);
+    }
+
+    /// Insert (key, value) pair in the hash map, it may grow one bucket if current load factor exceeds the threshold.
+    /// Note it may not split the actual overflowed bucket.
+    /// Abort if `key` already exists.
+    public fun insert<K, V>(map: &mut HashMap<K, V>, key: K, value: V) {
+        let hash = sip_hash(&key);
+        let index = bucket_index(map.level, map.num_buckets, hash);
+        let bucket = Table::borrow_mut(&mut map.buckets, &index);
+        let i = 0;
+        let len = Vector::length(bucket);
+        while (i < len) {
+            let entry = Vector::borrow(bucket, i);
+            assert!(&entry.key != &key, 0);
+            i = i + 1;
+        };
+        Vector::push_back(bucket, Entry {hash, key, value});
+        map.len = map.len + 1;
+
+        if (load_factor(map) > SPLIT_THRESHOLD) {
```

Comment:
Should the threshold be configurable in the ctor?

Reply:
I read a few hashmap implementation from different libraries, I don't see any of them expose the threshold to developers

Reply:
Java, the mother of all frameworks, has: https://docs.oracle.com/javase/8/docs/api/java/util/HashMap.html#HashMap-int-float-

However I'd agree its not that important.

Reply:
if we can support overloaded constructors or default values 🤷‍♂️